File: skills/docforge/scripts/manifest_deps.js
#!/usr/bin/env node
"use strict";
// Structured dependency extraction from project-definition manifests.
// Node peer of manifest_deps.py — identical algorithm so detect_profiles output
// stays byte-identical across runtimes. Built-in modules only.

const fs = require("fs");
const path = require("path");

const MAX_MANIFEST_BYTES = 1024 * 1024;
const ECOSYSTEMS = ["npm", "composer", "pip", "cargo", "go", "gem", "maven", "gradle", "nuget", "pub"];

function compareText(a, b) {
  return Buffer.compare(Buffer.from(a, "utf8"), Buffer.from(b, "utf8"));
}

function normalize(ecosystem, name) {
  let value = name.trim().toLowerCase();
  if (ecosystem === "pip") {
    let collapsed = "";
    let previousDash = false;
    for (const char of value) {
      if (char === "-" || char === "_" || char === ".") {
        if (!previousDash) collapsed += "-";
        previousDash = true;
      } else {
        collapsed += char;
        previousDash = false;
      }
    }
    value = collapsed.replace(/^-+/, "").replace(/-+$/, "");
  }
  return value;
}

function read(target) {
  try {
    if (fs.statSync(target).size > MAX_MANIFEST_BYTES) return null;
    return fs.readFileSync(target, "utf8");
  } catch (error) {
    return null;
  }
}

function jsonKeys(text, sections) {
  const names = [];
  let data;
  try {
    data = JSON.parse(text);
  } catch (error) {
    return names;
  }
  if (typeof data !== "object" || data === null || Array.isArray(data)) return names;
  for (const section of sections) {
    const table = data[section];
    if (table && typeof table === "object" && !Array.isArray(table)) {
      for (const key of Object.keys(table)) names.push(String(key));
    }
  }
  return names;
}

function isNameChar(char) {
  return /[A-Za-z0-9]/.test(char) || char === "-" || char === "_" || char === ".";
}

function pep508Name(spec) {
  let token = spec.trim();
  for (const cut of [";", " ", "\t"]) {
    const index = token.indexOf(cut);
    if (index >= 0) token = token.slice(0, index);
  }
  let stop = token.length;
  for (let index = 0; index < token.length; index++) {
    if (!isNameChar(token[index])) {
      stop = index;
      break;
    }
  }
  return token.slice(0, stop);
}

function requirements(text) {
  const names = [];
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.split("#")[0].trim();
    if (!line || line.startsWith("-")) continue;
    const name = pep508Name(line);
    if (name) names.push(name);
  }
  return names;
}

function sectionHeader(line) {
  const stripped = line.trim();
  if (stripped.startsWith("[") && stripped.endsWith("]") && stripped.length > 2) {
    return stripped.slice(1, -1).trim().replace(/^["']|["']$/g, "");
  }
  return null;
}

function quotedTokens(text) {
  const tokens = [];
  let quote = "";
  let current = "";
  for (const char of text) {
    if (quote) {
      if (char === quote) {
        tokens.push(current);
        current = "";
        quote = "";
      } else {
        current += char;
      }
    } else if (char === '"' || char === "'") {
      quote = char;
    }
  }
  return tokens;
}

function pyproject(text) {
  const names = [];
  let section = "";
  let inArray = false;
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.split("#")[0];
    const header = sectionHeader(line);
    if (header !== null) {
      section = header;
      inArray = false;
      continue;
    }
    let stripped = line.trim();
    if (!stripped) continue;
    if (section === "project") {
      const key = stripped.split("=")[0].trim();
      if (inArray || key === "dependencies") {
        if (stripped.includes("=") && key === "dependencies") {
          stripped = stripped.slice(stripped.indexOf("=") + 1);
        }
        inArray = !stripped.includes("]");
        for (const spec of quotedTokens(stripped)) {
          const name = pep508Name(spec);
          if (name) names.push(name);
        }
      }
    } else if (section === "project.optional-dependencies") {
      const body = stripped.includes("=") ? stripped.slice(stripped.indexOf("=") + 1) : stripped;
      for (const spec of quotedTokens(body)) {
        const name = pep508Name(spec);
        if (name) names.push(name);
      }
    } else if (section.startsWith("tool.poetry") && section.endsWith("dependencies")) {
      const key = stripped.split("=")[0].trim().replace(/^["']|["']$/g, "");
      if (key && key.toLowerCase() !== "python") names.push(key);
    }
  }
  return names;
}

function cargo(text) {
  const names = [];
  let section = "";
  const deps = new Set(["dependencies", "dev-dependencies", "build-dependencies"]);
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.split("#")[0];
    const header = sectionHeader(line);
    if (header !== null) {
      section = header;
      const parts = header.split(".");
      for (let index = 0; index < parts.length; index++) {
        if (deps.has(parts[index]) && index + 1 < parts.length) {
          const candidate = parts[index + 1];
          if (candidate && !candidate.startsWith("'") && !candidate.startsWith('"')) names.push(candidate);
          break;
        }
      }
      continue;
    }
    const stripped = line.trim();
    const leaf = section.split(".").pop();
    if (!stripped || !deps.has(leaf)) continue;
    const key = stripped.split("=")[0].split(".")[0].trim().replace(/^["']|["']$/g, "");
    if (key) names.push(key);
  }
  return names;
}

function goMod(text) {
  const names = [];
  let inBlock = false;
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.split("//")[0].trim();
    if (!line) continue;
    if (inBlock) {
      if (line.startsWith(")")) {
        inBlock = false;
        continue;
      }
      names.push(line.split(/\s+/)[0]);
    } else if (line.startsWith("require (")) {
      inBlock = true;
    } else if (line.startsWith("require ")) {
      const parts = line.slice("require ".length).trim().split(/\s+/);
      if (parts.length && parts[0]) names.push(parts[0]);
    }
  }
  return names;
}

function gemfile(text) {
  const names = [];
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.split("#")[0].trim();
    if (!line.startsWith("gem ")) continue;
    const tokens = quotedTokens(line);
    if (tokens.length) names.push(tokens[0]);
  }
  return names;
}

function innerText(block, tag) {
  const openTag = "<" + tag;
  const start = block.indexOf(openTag);
  if (start < 0) return null;
  const gt = block.indexOf(">", start);
  if (gt < 0) return null;
  const end = block.indexOf("</" + tag, gt);
  if (end < 0) return null;
  return block.slice(gt + 1, end).trim();
}

function blocks(text, tag) {
  const found = [];
  const openTag = "<" + tag;
  const closeTag = "</" + tag + ">";
  let search = 0;
  while (true) {
    const start = text.indexOf(openTag, search);
    if (start < 0) break;
    const end = text.indexOf(closeTag, start);
    if (end < 0) break;
    found.push(text.slice(start, end + closeTag.length));
    search = end + closeTag.length;
  }
  return found;
}

function pom(text) {
  const names = [];
  for (const tag of ["dependency", "plugin"]) {
    for (const block of blocks(text, tag)) {
      const group = innerText(block, "groupId");
      const artifact = innerText(block, "artifactId");
      if (group) names.push(group);
      if (group && artifact) names.push(group + ":" + artifact);
    }
  }
  return names;
}

function attr(text, name) {
  const values = [];
  const needle = name + "=";
  let search = 0;
  while (true) {
    const index = text.indexOf(needle, search);
    if (index < 0) break;
    const rest = text.slice(index + needle.length).replace(/^\s+/, "");
    if (rest && (rest[0] === '"' || rest[0] === "'")) {
      const quote = rest[0];
      const end = rest.indexOf(quote, 1);
      if (end > 0) {
        values.push(rest.slice(1, end));
        search = index + needle.length + end + 1;
        continue;
      }
    }
    search = index + needle.length;
  }
  return values;
}

function csproj(text) {
  const names = [];
  for (const value of attr(text, "Sdk")) names.push(value);
  let search = 0;
  while (true) {
    const index = text.indexOf("<PackageReference", search);
    if (index < 0) break;
    const gt = text.indexOf(">", index);
    const segment = text.slice(index, gt > 0 ? gt : text.length);
    for (const value of attr(segment, "Include")) names.push(value);
    search = gt > 0 ? gt + 1 : text.length;
  }
  return names;
}

function gradle(text) {
  const names = [];
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.split("//")[0].trim();
    if (!line) continue;
    const tokens = quotedTokens(line);
    const words = line.replace(/\(/g, " ").replace(/\)/g, " ").split(/\s+/);
    if (words.includes("id") || line.includes("plugin") || line.includes("plugins")) {
      for (const token of tokens) names.push(token);
    }
    for (const token of tokens) {
      if (token.includes(":")) {
        const parts = token.split(":");
        if (parts.length >= 2 && parts[0] && parts[1]) {
          names.push(parts[0] + ":" + parts[1]);
          names.push(parts[0]);
        }
      }
    }
  }
  return names;
}

function pubspec(text) {
  const names = [];
  let inDeps = false;
  for (const raw of text.split(/\r?\n/)) {
    if (!raw.trim() || raw.replace(/^\s+/, "").startsWith("#")) continue;
    const indent = raw.length - raw.replace(/^\s+/, "").length;
    const stripped = raw.trim();
    if (indent === 0) {
      const key = stripped.split(":")[0].trim();
      inDeps = key === "dependencies" || key === "dev_dependencies";
      continue;
    }
    if (inDeps && indent <= 2 && stripped.includes(":")) {
      const key = stripped.split(":")[0].trim();
      if (key) names.push(key);
    }
  }
  return names;
}

function classify(name) {
  const lower = name.toLowerCase();
  if (name === "package.json") return ["npm", "npm"];
  if (name === "composer.json") return ["composer", "composer"];
  if (lower.startsWith("requirements") && lower.endsWith(".txt")) return ["pip", "requirements"];
  if (name === "pyproject.toml") return ["pip", "pyproject"];
  if (name === "Cargo.toml") return ["cargo", "cargo"];
  if (name === "go.mod") return ["go", "go"];
  if (name === "Gemfile") return ["gem", "gem"];
  if (name === "pom.xml") return ["maven", "maven"];
  if (name.endsWith(".gradle") || name.endsWith(".gradle.kts")) return ["gradle", "gradle"];
  if (name.endsWith(".csproj")) return ["nuget", "csproj"];
  if (name === "pubspec.yaml" || name === "pubspec.yml") return ["pub", "pubspec"];
  return null;
}

const PARSERS = {
  npm: (text) => jsonKeys(text, ["dependencies", "devDependencies", "peerDependencies", "optionalDependencies"]),
  composer: (text) => jsonKeys(text, ["require", "require-dev"]),
  requirements,
  pyproject,
  cargo,
  go: goMod,
  gem: gemfile,
  maven: pom,
  csproj,
  gradle,
  pubspec,
};

function extractDependencies(files) {
  const index = {};
  for (const [relative, target] of files) {
    const name = path.posix.basename(relative.split(path.sep).join("/"));
    const classified = classify(name);
    if (classified === null) continue;
    const [ecosystem, parserKey] = classified;
    const text = read(target);
    if (text === null) continue;
    for (const rawName of PARSERS[parserKey](text)) {
      const key = normalize(ecosystem, rawName);
      if (!key) continue;
      if (!index[ecosystem]) index[ecosystem] = {};
      if (!index[ecosystem][key]) index[ecosystem][key] = [];
      if (!index[ecosystem][key].includes(relative)) index[ecosystem][key].push(relative);
    }
  }
  for (const ecosystem of Object.keys(index)) {
    for (const key of Object.keys(index[ecosystem])) {
      index[ecosystem][key] = [...new Set(index[ecosystem][key])].sort(compareText);
    }
  }
  return index;
}

module.exports = { extractDependencies, normalize, ECOSYSTEMS };

File: skills/docforge/scripts/manifest_deps.py
#!/usr/bin/env python3
"""Structured dependency extraction from project-definition manifests.

Shared helper for detect_profiles: turns the file inventory into a per-ecosystem
index of declared dependency identifiers, each mapped to the manifest paths that
declared it. Detection matches a `dependency` signal against this index instead
of substring-scanning raw text, so `torch` no longer bleeds to `torchvision` and
`Django` no longer depends on letter case.

Built-in-only. The Node peer (`manifest_deps.js`) runs the identical algorithm so
`detect_profiles` output stays byte-identical across runtimes. The correctness
bar for the TOML/XML/YAML readers is "extract the declared dependency names",
not "be a conformant parser".
"""

from __future__ import annotations

import json
from pathlib import Path

MAX_MANIFEST_BYTES = 1024 * 1024

ECOSYSTEMS = ["npm", "composer", "pip", "cargo", "go", "gem", "maven", "gradle", "nuget", "pub"]


def normalize(ecosystem: str, name: str) -> str:
    value = name.strip().lower()
    if ecosystem == "pip":
        collapsed = []
        previous_dash = False
        for char in value:
            if char in "-_.":
                if not previous_dash:
                    collapsed.append("-")
                previous_dash = True
            else:
                collapsed.append(char)
                previous_dash = False
        value = "".join(collapsed).strip("-")
    return value


def _read(path: Path) -> str | None:
    try:
        if path.stat().st_size > MAX_MANIFEST_BYTES:
            return None
        return path.read_text(encoding="utf-8", errors="replace")
    except OSError:
        return None


def _json_keys(text: str, sections: list[str]) -> list[str]:
    names: list[str] = []
    try:
        data = json.loads(text)
    except (ValueError, TypeError):
        return names
    if not isinstance(data, dict):
        return names
    for section in sections:
        table = data.get(section)
        if isinstance(table, dict):
            names.extend(str(key) for key in table.keys())
    return names


def _pep508_name(spec: str) -> str:
    token = spec.strip()
    for cut in (";", " ", "\t"):
        index = token.find(cut)
        if index >= 0:
            token = token[:index]
    stop = len(token)
    for index, char in enumerate(token):
        if not (char.isalnum() or char in "-_."):
            stop = index
            break
    return token[:stop]


def _requirements(text: str) -> list[str]:
    names: list[str] = []
    for raw in text.splitlines():
        line = raw.split("#", 1)[0].strip()
        if not line or line.startswith("-"):
            continue
        name = _pep508_name(line)
        if name:
            names.append(name)
    return names


def _section_header(line: str) -> str | None:
    stripped = line.strip()
    if stripped.startswith("[") and stripped.endswith("]") and len(stripped) > 2:
        return stripped[1:-1].strip().strip('"').strip("'")
    return None


def _quoted_tokens(text: str) -> list[str]:
    tokens: list[str] = []
    quote = ""
    current: list[str] = []
    for char in text:
        if quote:
            if char == quote:
                tokens.append("".join(current))
                current = []
                quote = ""
            else:
                current.append(char)
        elif char in "\"'":
            quote = char
    return tokens


def _pyproject(text: str) -> list[str]:
    names: list[str] = []
    section = ""
    in_array = False
    for raw in text.splitlines():
        line = raw.split("#", 1)[0]
        header = _section_header(line)
        if header is not None:
            section = header
            in_array = False
            continue
        stripped = line.strip()
        if not stripped:
            continue
        if section == "project":
            key = stripped.split("=", 1)[0].strip()
            if in_array or key == "dependencies":
                if "=" in stripped and key == "dependencies":
                    stripped = stripped.split("=", 1)[1]
                in_array = "]" not in stripped
                for spec in _quoted_tokens(stripped):
                    name = _pep508_name(spec)
                    if name:
                        names.append(name)
        elif section == "project.optional-dependencies":
            body = stripped.split("=", 1)[1] if "=" in stripped else stripped
            for spec in _quoted_tokens(body):
                name = _pep508_name(spec)
                if name:
                    names.append(name)
        elif section.startswith("tool.poetry") and section.endswith("dependencies"):
            key = stripped.split("=", 1)[0].strip().strip('"').strip("'")
            if key and key.lower() != "python":
                names.append(key)
    return names


def _cargo(text: str) -> list[str]:
    names: list[str] = []
    section = ""
    deps = {"dependencies", "dev-dependencies", "build-dependencies"}
    for raw in text.splitlines():
        line = raw.split("#", 1)[0]
        header = _section_header(line)
        if header is not None:
            section = header
            parts = header.split(".")
            for index, part in enumerate(parts):
                if part in deps and index + 1 < len(parts):
                    candidate = parts[index + 1]
                    if candidate and not candidate.startswith("'") and not candidate.startswith('"'):
                        names.append(candidate)
                    break
            continue
        stripped = line.strip()
        if not stripped or section.split(".")[-1] not in deps:
            continue
        key = stripped.split("=", 1)[0].split(".", 1)[0].strip().strip('"').strip("'")
        if key:
            names.append(key)
    return names


def _go_mod(text: str) -> list[str]:
    names: list[str] = []
    in_block = False
    for raw in text.splitlines():
        line = raw.split("//", 1)[0].strip()
        if not line:
            continue
        if in_block:
            if line.startswith(")"):
                in_block = False
                continue
            token = line.split()[0]
            names.append(token)
        elif line.startswith("require ("):
            in_block = True
        elif line.startswith("require "):
            parts = line[len("require "):].strip().split()
            if parts:
                names.append(parts[0])
    return names


def _gemfile(text: str) -> list[str]:
    names: list[str] = []
    for raw in text.splitlines():
        line = raw.split("#", 1)[0].strip()
        if not line.startswith("gem "):
            continue
        tokens = _quoted_tokens(line)
        if tokens:
            names.append(tokens[0])
    return names


def _inner_text(block: str, tag: str) -> str | None:
    open_tag = "<" + tag
    start = block.find(open_tag)
    if start < 0:
        return None
    gt = block.find(">", start)
    if gt < 0:
        return None
    end = block.find("</" + tag, gt)
    if end < 0:
        return None
    return block[gt + 1:end].strip()


def _blocks(text: str, tag: str) -> list[str]:
    blocks: list[str] = []
    open_tag = "<" + tag
    close_tag = "</" + tag + ">"
    search = 0
    while True:
        start = text.find(open_tag, search)
        if start < 0:
            break
        end = text.find(close_tag, start)
        if end < 0:
            break
        blocks.append(text[start:end + len(close_tag)])
        search = end + len(close_tag)
    return blocks


def _pom(text: str) -> list[str]:
    names: list[str] = []
    for tag in ("dependency", "plugin"):
        for block in _blocks(text, tag):
            group = _inner_text(block, "groupId")
            artifact = _inner_text(block, "artifactId")
            if group:
                names.append(group)
            if group and artifact:
                names.append(group + ":" + artifact)
    return names


def _attr(text: str, attr: str) -> list[str]:
    values: list[str] = []
    needle = attr + "="
    search = 0
    while True:
        index = text.find(needle, search)
        if index < 0:
            break
        rest = text[index + len(needle):].lstrip()
        if rest and rest[0] in "\"'":
            quote = rest[0]
            end = rest.find(quote, 1)
            if end > 0:
                values.append(rest[1:end])
                search = index + len(needle) + end + 1
                continue
        search = index + len(needle)
    return values


def _csproj(text: str) -> list[str]:
    names: list[str] = []
    for value in _attr(text, "Sdk"):
        names.append(value)
    # PackageReference is usually self-closing; scan Include attributes on it.
    search = 0
    while True:
        index = text.find("<PackageReference", search)
        if index < 0:
            break
        gt = text.find(">", index)
        segment = text[index:gt if gt > 0 else len(text)]
        for value in _attr(segment, "Include"):
            names.append(value)
        search = gt + 1 if gt > 0 else len(text)
    return names


def _gradle(text: str) -> list[str]:
    names: list[str] = []
    for raw in text.splitlines():
        line = raw.split("//", 1)[0].strip()
        if not line:
            continue
        tokens = _quoted_tokens(line)
        words = line.replace("(", " ").replace(")", " ").split()
        if "id" in words or "plugin" in line or "plugins" in line:
            for token in tokens:
                names.append(token)
        for token in tokens:
            if ":" in token:
                parts = token.split(":")
                if len(parts) >= 2 and parts[0] and parts[1]:
                    names.append(parts[0] + ":" + parts[1])
                    names.append(parts[0])
    return names


def _pubspec(text: str) -> list[str]:
    names: list[str] = []
    in_deps = False
    for raw in text.splitlines():
        if not raw.strip() or raw.lstrip().startswith("#"):
            continue
        indent = len(raw) - len(raw.lstrip())
        stripped = raw.strip()
        if indent == 0:
            key = stripped.split(":", 1)[0].strip()
            in_deps = key in {"dependencies", "dev_dependencies"}
            continue
        if in_deps and indent <= 2 and ":" in stripped:
            key = stripped.split(":", 1)[0].strip()
            if key:
                names.append(key)
    return names


def _classify(relative: str, name: str) -> tuple[str, str] | None:
    lower = name.lower()
    if name == "package.json":
        return ("npm", "npm")
    if name == "composer.json":
        return ("composer", "composer")
    if lower.startswith("requirements") and lower.endswith(".txt"):
        return ("pip", "requirements")
    if name == "pyproject.toml":
        return ("pip", "pyproject")
    if name == "Cargo.toml":
        return ("cargo", "cargo")
    if name == "go.mod":
        return ("go", "go")
    if name == "Gemfile":
        return ("gem", "gem")
    if name == "pom.xml":
        return ("maven", "maven")
    if name.endswith(".gradle") or name.endswith(".gradle.kts"):
        return ("gradle", "gradle")
    if name.endswith(".csproj"):
        return ("nuget", "csproj")
    if name in {"pubspec.yaml", "pubspec.yml"}:
        return ("pub", "pubspec")
    return None


_PARSERS = {
    "npm": lambda text: _json_keys(text, ["dependencies", "devDependencies", "peerDependencies", "optionalDependencies"]),
    "composer": lambda text: _json_keys(text, ["require", "require-dev"]),
    "requirements": _requirements,
    "pyproject": _pyproject,
    "cargo": _cargo,
    "go": _go_mod,
    "gem": _gemfile,
    "maven": _pom,
    "csproj": _csproj,
    "gradle": _gradle,
    "pubspec": _pubspec,
}


def extract_dependencies(files: list[tuple[str, Path]]) -> dict:
    """Return {ecosystem: {normalized_name: [manifest_paths]}} from the inventory."""
    index: dict = {}
    for relative, path in files:
        name = Path(relative).name
        classified = _classify(relative, name)
        if classified is None:
            continue
        ecosystem, parser_key = classified
        text = _read(path)
        if text is None:
            continue
        for raw_name in _PARSERS[parser_key](text):
            key = normalize(ecosystem, raw_name)
            if not key:
                continue
            bucket = index.setdefault(ecosystem, {})
            paths = bucket.setdefault(key, [])
            if relative not in paths:
                paths.append(relative)
    for ecosystem in index:
        for key in index[ecosystem]:
            index[ecosystem][key] = sorted(set(index[ecosystem][key]))
    return index
